flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -0,0 +1,296 @@
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.jobmaster.LogicalSlot;
+import org.apache.flink.runtime.shuffle.ShuffleDescriptor;
+import org.apache.flink.runtime.shuffle.UnknownShuffleDescriptor;
+import org.apache.flink.types.Either;
+import org.apache.flink.util.SerializedValue;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Factory of {@link TaskDeploymentDescriptor} to deploy {@link org.apache.flink.runtime.taskmanager.Task} from {@link Execution}.
+ */
+public class TaskDeploymentDescriptorFactory {
+	private final ExecutionAttemptID executionId;
+	private final int attemptNumber;
+	private final MaybeOffloaded<JobInformation> serializedJobInformation;
+	private final MaybeOffloaded<TaskInformation> taskInfo;
+	private final JobID jobID;
+	private final boolean lazyScheduling;
+	private final int subtaskIndex;
+	private final ExecutionEdge[][] inputEdges;
+
+	private TaskDeploymentDescriptorFactory(
+			ExecutionAttemptID executionId,
+			int attemptNumber,
+			MaybeOffloaded<JobInformation> serializedJobInformation,
+			MaybeOffloaded<TaskInformation> taskInfo,
+			JobID jobID,
+			boolean lazyScheduling,
+			int subtaskIndex,
+			ExecutionEdge[][] inputEdges) {
+		this.executionId = executionId;
+		this.attemptNumber = attemptNumber;
+		this.serializedJobInformation = serializedJobInformation;
+		this.taskInfo = taskInfo;
+		this.jobID = jobID;
+		this.lazyScheduling = lazyScheduling;
+		this.subtaskIndex = subtaskIndex;
+		this.inputEdges = inputEdges;
+	}
+
+	public TaskDeploymentDescriptor createDeploymentDescriptor(
+			LogicalSlot targetSlot,
+			@Nullable JobManagerTaskRestore taskRestore,
+			Collection<ResultPartitionDeploymentDescriptor> producedPartitions) throws Exception {
+		return createDeploymentDescriptor(
+			targetSlot.getTaskManagerLocation().getResourceID(),
+			targetSlot.getAllocationId(),
+			targetSlot.getPhysicalSlotNumber(),
+			taskRestore,
+			producedPartitions);
+	}
+
+	@VisibleForTesting
+	public TaskDeploymentDescriptor createDeploymentDescriptor(
+			ResourceID location,
+			AllocationID allocationID,
+			int targetSlotNumber,
+			@Nullable JobManagerTaskRestore taskRestore,
+			Collection<ResultPartitionDeploymentDescriptor> producedPartitions) throws Exception {
+		return new TaskDeploymentDescriptor(
+			jobID,
+			serializedJobInformation,
+			taskInfo,
+			executionId,
+			allocationID,
+			subtaskIndex,
+			attemptNumber,
+			targetSlotNumber,
+			taskRestore,
+			new ArrayList<>(producedPartitions),
+			createInputGateDeploymentDescriptors(location, inputEdges, subtaskIndex, lazyScheduling));
```

Comment:
I try to understand the thoughts of which infos should be passed to construct the factory and which infos should be passed in `createDeploymentDescriptor` to create TDD.
I think it is reasonable to pass `ExecutionVertex` to construct the factory as above, and we could get all the necessary topology infos from vertex. And during `createDeploymentDescriptor`, it seems easy to understand if we create both IGDD and RPDD here. Now RPDD was created during assigning slot before. Actually creation of IGDD and RPDD are both relying on infos in slot, so RPDD could also be created together with IGDD here. Or I missed something?